### PR TITLE
dist: bump `rustup` version to `1.27.0`

### PR DESCRIPTION
**Pull request by rami3l**
cc #3501.

## Concerns

- [x] Should we make a beta-driving phase for this?

---

# Changelog

## [1.27.0] - 2024-03-08

This long-awaited Rustup release has gathered all the new features and fixes since April 2023.
These changes include improvements in Rustup's maintainability, user experience, compatibility and documentation quality.

The headlines of this release are:
- Basic support for `fish` shell has been added.
- Support for the `loongarch64-unknown-linux-gnu` host platform has been added.

Also, it's worth mentioning that Dirkjan Ochtman and rami3l have joined the team and are coordinating this new release.

Finally, the project seems to have attracted a total of 23 new contributors within this release cycle. Looking forward to seeing you again in the future!

### Added

- Add basic support for `fish` shell [pr#3108] 
- Add the `RUSTUP_TERM_COLOR` environment variable to force the use of colored output [pr#3435]
- Improve `rustup-init.sh`'s compatibility with `ksh` and `zsh` [pr#3475]
- Add a warning when running under Rosetta 2 [pr#3068]
- Add browser detection for RISC-V 64 platform [pr#3642]
- Add a warning when removing the last/host target for a toolchain [pr#3637]

### Changed

- Upgrade `clap` to v4 [pr#3444]
- Fix incorrect platform detection on macOS aarch64 due to Rosetta 2 [pr#3438]
- Fix incorrect platform detection on 32-bit Linux userland with a 64-bit kernel [pr#3488] [pr#3490]
- Improve Windows system32 DLL loading mechanism [pr#3493]
- Improve suggestions about missing components [pr#3453]
- Fix handling of toolchain names with special characters [pr#3518]
- Fix panic in `component list --toolchain stable` [pr#3548]
- Rename `llvm-tools-preview` component to `llvm-tools` [pr#3578]
- Try using symlinks if possible on Windows [pr#3687]
- Bump a lot of dependencies to their latest versions [pr#renovate-bot]

Thanks go to:

- Anthony Perkins (acperkins)
- Tianqi (airstone42)
- Alex Gaynor (alex)
- Alex Hudspith (alexhudspith)
- Alan Somers (asomers)
- Brett (brettearle)
- Burak Emir (burakemir)
- Chris Denton (ChrisDenton)
- cui fliter (cuishuang)
- Dirkjan Ochtman (djc)
- Dezhi Wu (dzvon)
- Eric Swanson (ericswanson-dfinity)
- Prikshit Gautam (gautamprikshit1)
- hev (heiher)
- 二手掉包工程师 (hi-rustin)
- Kamila Borowska (KamilaBorowska)
- klensy (klensy)
- Jakub Beránek (Kobzol)
- Kornel (kornelski)
- Matt Harding (majaha)
- Mathias Brossard (mbrossard)
- Christian Thackston (nan60)
- Ruohui Wang (noirgif)
- Olivier Lemasle (olivierlemasle)
- Chih Wang (ongchi)
- Pavel Roskin (proski)
- rami3l (rami3l)
- Robert Collins (rbtcollins)
- Sandesh  Pyakurel (Sandesh-Pyakurel)
- Waffle Maybe (WaffleLapkin)
- Jubilee (workingjubilee)
- WÁNG Xuěruì (xen0n)
- Yerkebulan Tulibergenov (yerke)
- Renovate Bot (renovate)

**Full Changelog**: https://github.com/rust-lang/rustup/compare/1.26.0...1.27.0

[1.27.0]: https://github.com/rust-lang/rustup/releases/tag/1.27.0
[pr#3108]: https://github.com/rust-lang/rustup/pull/3108
[pr#3435]: https://github.com/rust-lang/rustup/pull/3435
[pr#3475]: https://github.com/rust-lang/rustup/pull/3475
[pr#3068]: https://github.com/rust-lang/rustup/pull/3068
[pr#3642]: https://github.com/rust-lang/rustup/pull/3642
[pr#3637]: https://github.com/rust-lang/rustup/pull/3637
[pr#3444]: https://github.com/rust-lang/rustup/pull/3444
[pr#3438]: https://github.com/rust-lang/rustup/pull/3438
[pr#3488]: https://github.com/rust-lang/rustup/pull/3488
[pr#3490]: https://github.com/rust-lang/rustup/pull/3490
[pr#3493]: https://github.com/rust-lang/rustup/pull/3493
[pr#3453]: https://github.com/rust-lang/rustup/pull/3453
[pr#3518]: https://github.com/rust-lang/rustup/pull/3518
[pr#3548]: https://github.com/rust-lang/rustup/pull/3548
[pr#3578]: https://github.com/rust-lang/rustup/pull/3578
[pr#3687]: https://github.com/rust-lang/rustup/pull/3687
[pr#renovate-bot]: https://github.com/rust-lang/rustup/pulls?q=is:pr+author:app/renovate+is:merged

### CODE REVIEW

**Comment by rami3l** [comment]
Oops that changelog is way too long. I have to truncate it a bit...

**Comment by 0xPoe** [comment]
Thanks for working on this! https://rust-lang.github.io/rustup/dev-guide/release-process.html Just a reminder: 
1. We should merge the master's changes to stable.
2. We need to bump the versions and add a changelog on stable. (I usually create a PR target to stable, once all CI is green, I will push it from local, and then GitHub will automatically close the PR).
3. After we ask the t-release team to deliver the binaries, then we can merge the stable branch back to the master branch. (version bump + changelogs will merge back).

I guess the point is we need to release it based on the stable branch.

Thanks again! Very excited to see rustup `1.27.0`.

**Comment by rami3l** [comment]
@hi-rustin Oops, looks like now stable and master are (and might always be) one commit off?

image

... or is it possible to force-push master to stable?

(cc @kinnison)

**Comment by 0xPoe** [comment]
@rami3l Can we merge that commit back to master?

**Comment by rami3l** [comment]
> @rami3l Can we merge that commit back to master?

@hi-rustin I guess that might not be easy since I've made #3491 when I wasn't aware of the stable branch...

**Comment by 0xPoe** [comment]
> > @rami3l Can we merge that commit back to master?
> 
> @hi-rustin I guess that might not be easy since I've made #3491 when I wasn't aware of the stable branch...

How about we revert this PR? Then maybe we can merge stable back? (I am not very sure). We need to do some tests in our local repo.

**Comment by rami3l** [comment]
> How about we revert this PR? Then maybe we can merge stable back? (I am not very sure). We need to do some tests in our local repo.

@hi-rustin Rather, I meant that I'm not very sure if our branch protection rules allow me to force-push on `stable`. According to my understanding it will work if it's allowed by the organization... That being said, don't worry. I believe we can find a solution.

---

Update: Just asked infra, it turns out there are currently no restrictions on that. I'll try that then. (Sorry for overwriting your commit!)

**Comment by kinnison** [comment]
Generally speaking (a) you should always merge back to master whatever ended up on stable after a release and (b) if stable does end up out-of-sync you need to *carefully* unpick what/why because it's the commit SHAs on stable which end up in release binaries so you don't want to lose them if at all possible.

**Comment by kinnison** [comment]
Otherwise, congratulations on this release, it looks like an excellent combination of features and fixes.

**Comment by rami3l** [comment]
> Generally speaking (a) you should always merge back to master whatever ended up on stable after a release and (b) if stable does end up out-of-sync you need to _carefully_ unpick what/why because it's the commit SHAs on stable which end up in release binaries so you don't want to lose them if at all possible.

@kinnison Thanks for the info! I checked again and I believe this change has been covered by https://github.com/rust-lang/rustup/pull/3491.

https://github.com/rust-lang/rustup/compare/dist/release-1-27...stable gives:

image

Also, according to the current diff, the latest stable checksum to date is `577bf51ae 2023-04-05` which I believe is unrelated to this change made one month later.

**Comment by 0xPoe** [comment]
@rami3l I think there are two solutions for this issue.
1. Force push master to the stable. But we need to make sure we don't change any commit SHA before the 1.26.0 tag. 
2. We can try to revert https://github.com/rust-lang/rustup/pull/3491, and then try to merge stable to the master first(Need some local repo tests before we do it). Then we can merge the master to stable and start the release process on the stable branch.

I will run some quick tests for the second solution. If it works, then we won't need to do a force push.

**Comment by 0xPoe** [comment]
I tested soultion2 in my forked repo:
1. revert the PR 3491 in this PR: https://github.com/hi-rustin/rustup/pull/1
2. Create a PR from stable to master: https://github.com/hi-rustin/rustup/pull/2
3. Create a PR from master to stable: https://github.com/hi-rustin/rustup/pull/3

It seemed to work well. Sorry again for forgetting to merge that commit back.

I will let the team make the decision! Thanks again for working on this!

**Comment by rami3l** [comment]
> I tested soultion2 in my forked repo:
> ...
> It seemed to work well. Sorry again for forgetting to merge that commit back.
> 
> I will let the team make the decision! Thanks again for working on this!

@hi-rustin Thanks again for trying it out!

However it still looks problematic in our case since GitHub Merge Queue is in action now and I guess that might change the shasums on step 2 anyway (unless I can still merge manually just for the release, I'm not quite sure about that). We have full control over the `stable` branch though so we can be sure that it is indeed in sync with `master`, but not in the opposite direction. Further discussions with infra might be required to move this forward... 🙇

**Comment by 0xPoe** [comment]
> However it still looks problematic in our case since GitHub Merge Queue is in action now and I guess that might change the shasums on step 2 anyway (unless I can still merge manually just for the release, I'm not quite sure about that).

I see, we are using the rebase merge method now. But I guess this would also cause a problem when we want to merge the stable back to the master after we release 1.27.0.

**Comment by rami3l** [comment]
> I see, we are using the rebase merge method now. But I guess this would also cause a problem when we want to merge the stable back to the master after we release 1.27.0.

Exactly, so IMHO we shouldn't. More precisely, from now on `master` should always be ahead of `stable`, and we can sync `stable` with `master` from time to time.

The only subtlety remaining seems to be the commit shasum in the release:

> ... we seem to have two different ones for the last release already:
> 1. The shasum in `rustup.exe --help`: `5af9b9484 2023-04-05`
> 2. The shasum in `rustup-init.sh --help`: `577bf51ae 2023-04-05`

In conclusion, with the new merge-queue-enabled workflow, if we still do the release in one PR:
- 1 and 2 still exist and remain traceable, and they will be different as in 1.26.
- 2 won't be found on `stable`, since it's been rewritten when being merged to `master` (but still traceable just like https://github.com/rust-lang/rustup/commit/fb7b525).

There is even a way of making 2 accessible on `stable` as well, which is making two subsequent PRs, with the second pointing `rustup-init.sh --help` to a certain commit on `master`, but 1 and 2 will still be different in this case.
Better yet, we might be able to perform a fast-forward manually in order to prevent shasum changes if possible.

**Comment by 0xPoe** [comment]
> we seem to have two different ones for the last release already:

I believe this is by design. You can find it in https://rust-lang.github.io/rustup/dev-guide/release-process.html.

Because we need to update the Cargo.lock and push it first. After that, we can get the commit for updating the Cargo.lock. 

After we finally build the binary and release it, then we get the different commits for them.

For instance:
1. The 1.25.1 init.sh sha: https://github.com/rust-lang/rustup/blob/bb60b1e89cc01a682dd16f2854329030938d3fe3/rustup-init.sh#L28 rustup-init 1.25.1 (48d233f65 2022-07-12)
2. The 1.25.1 tag sha: https://github.com/rust-lang/rustup/commit/bb60b1e89cc01a682dd16f2854329030938d3fe3

**Comment by rami3l** [comment]
In that case, I guess we can try to initiate the release process.
First, this PR will be merged to `master`; then the changes will be reflected on `stable`.

Next, since this is the first release since such a long time, I propose doing a beta-testing phase by (quoting @kinnison)
> getting as far as the release would normally on the stable branch and the dev-static.rlo push from t-release; then you stop, and do the extra testing.

I expect to keep it like that without actually releasing for a few weeks.
Hopefully this will buy us some time in order to decide whether to revert https://github.com/rust-lang/rustup/pull/3578, for example.

Thus, I'm marking this PR as ready again.

**Comment by heiher** [comment]
I'm sorry, I just found that the LoongArch build is broken. I'm currently working on [fixing libc](https://github.com/rust-lang/libc/pull/3570). If possible, I really want to catch up with the `1.27.0` release, which will be the first version to support LoongArch.

**Comment by rami3l** [comment]
@heiher No worries! We don't currently have a specific timeline here and even if the changes do get pushed to the `stable` branch, we can still include changes from `master` as long as the release is still not tagged and announced AFAIK.

File: Cargo.toml [hunk 1]
```diff
@@ -88,7 +88,9 @@ threadpool = "1"
 tokio = { workspace = true, optional = true }
 toml = "0.8"
 tracing-opentelemetry = { workspace = true, optional = true }
-tracing-subscriber = { workspace = true, optional = true, features = ["env-filter"] }
+tracing-subscriber = { workspace = true, optional = true, features = [
```
Comment (djc) [inline Cargo.toml]: Please undo the unrelated formatting changes?

File: rustup-init.sh [hunk 2]
```diff
@@ -31,7 +31,7 @@ RUSTUP_UPDATE_ROOT="${RUSTUP_UPDATE_ROOT:-https://static.rust-lang.org/rustup}"
 # NOTICE: If you change anything here, please make the same changes in setup_mode.rs
 usage() {
     cat <<EOF
-rustup-init 1.26.0 (577bf51ae 2023-04-05)
+rustup-init 1.27.0 (5623b7424 2024-01-29)
```
Comment (djc) [inline rustup-init.sh]: Presumably this is generated, so belongs in the previous commit? Or did you have to change this manually?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,75 @@
 # Changelog
 
+## [1.27.0] - 2024-01-29
+
+This long-awaited Rustup release has gathered all the new features and fixes since April 2023.
+These changes have greatly improved Rustup's maintainability, user experience, compatibility and documentation quality.
```
Comment (djc) [inline CHANGELOG.md]: I feel like this is overstating things a little? I don't feel like the Added/Changed lines below represent "great" improvements.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,75 @@
...
+
+The headlines of this release are:
+- Basic support for `fish` shell has been added.
+- Rustup now uses `clap` v4 for argument parsing.
```
Comment (djc) [inline CHANGELOG.md]: I would suggest we do not mention the clap upgrade and the GitHub merge queue changes here, since they aren't really relevant to our users.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,75 @@
...
+- Support and build test for `loongarch64-unknown-linux-gnu` has been added.
+- The project has introduced GitHub Merge Queue as part of its workflow, and the CI pipeline has been revamped accordingly.
+
+Also, the project seems to have attracted a substantial number of new contributors within this release cycle. Looking forward to seeing you again in the future!
```
Comment (djc) [inline CHANGELOG.md]: I would make this more factual, like "attracted 15 new contributors this release cycle". (Number to be checked, not sure where you got this?)

It's probably worth mentioning that there are two new team members coordinating this release?

Comment (rami3l) [reply CHANGELOG.md]: > I would make this more factual, like "attracted 15 new contributors this release cycle". (Number to be checked, not sure where you got this?)

GitHub can generate this, and I just checked, this time we have 21 actually!

> It's probably worth mentioning that there are two new team members coordinating this release?

Oh yup. Will do!

File: rustup-init.sh (hunk 2 above)
Comment (rami3l) [reply rustup-init.sh]: As per https://github.com/rust-lang/rustup/pull/3653#issuecomment-1918457423 I have to change this manually every time, unfortunately.
I don't like this myself either, but is there another way?
(cc @rbtcollins)

File: CHANGELOG.md (hunk 4 above)
Comment (rami3l) [reply CHANGELOG.md:10]: Will do.

File: CHANGELOG.md (hunk 3 above)
Comment (rami3l) [reply CHANGELOG.md:6]: @djc I guess you can directly edit the changelog on the top of this thread and I'll reflect the changes in the actual file, how does that sound?

File: rustup-init.sh (hunk 2 above)
Comment (rami3l) [reply rustup-init.sh]: The pain point here is that we don't know in advance what the commit shasum will be (and as this shasum will be changed with the merge queue, we might need an override somehow just for releasing or we should change the workflow).
The output of the script seems to be intentionally reflecting `rustup-init.exe` but the shasum will be always off-by-one as explained in the aforementioned comment.

File: CHANGELOG.md (hunk 4 above)
Comment (rami3l) [reply CHANGELOG.md]: Removed from the headlines. I do want to keep the clap v4 thing here since the CLI output is slightly changed. Although there's no clear evidence whether it is popular to parse Rustup's output, I think it's better to at least leave a hint.

File: rustup-init.sh (hunk 2 above)
Comment (rami3l) [reply rustup-init.sh]: I did some further research and I haven't found an ideal way of doing this automatically. I guess we should stick to the original plan then. First I'll try a fast-forwarding merge to bypass the merge queue. If that is disallowed in out repo, I'll issue another PR to fix up the commit shasum.

**Review by djc — APPROVED** [review]

**Comment by rami3l** [comment]
Note: As we have decided to include a beta-testing phase, merging this PR doesn't imply a new release (yet), so #3501 won't be closed for now.

However, we sincerely invite everyone in this thread to try out the beta by setting the following env variable:

```
RUSTUP_UPDATE_ROOT=https://dev-static.rust-lang.org/rustup
```

~~The actual beta release will be available shortly after the `stable` branch's CI turns green. Please stay tuned!~~ The beta release seems to be available right now. If you don't mind, please try it out and let us know what you think about it! :)

Here's the relevant internals forum thread: https://internals.rust-lang.org/t/seeking-beta-testers-for-rustup-1-27-0/20352

**Comment by rami3l** [comment]
@djc @rbtcollins Oops, the stable CI has failed. Considering the fact that it hasn't been executed in about a year, I guess that's also understandable. I'll do a quick diagnosis here and see what we can do:

- `i686-unknown-linux-gnu`: `Can't locate IPC/Cmd.pm in @INC`, looks like an OpenSSL-related issue previously known in another target's build environment, I'll backport the fix from https://github.com/rust-lang/rustup/pull/3478/commits/aa972e5abeecbad1189831fa5ef66eb9c385d001.
- `mips*-unknown-linux-gnu* x 4`: Failing as expected. They have been downgraded to tier 3 for which we should no longer ship precompiled binaries. I'll remove them.
- `i686-linux-android`: `ld: error: undefined symbol: __atomic_load`... (I have no idea but looks like an OpenSSL thing)
- `riscv64gc-unknown-linux-gnu`: `error: 'rv64i_zkne_*' undeclared (first use in this function);` (looks like an OpenSSL thing, probably https://github.com/alexcrichton/openssl-src-rs/issues/222#issuecomment-1917168867)

I have made #3668 hoping to address these concerns.

**Comment by rami3l** [comment]
The beta release seems to be available right now. If you don't mind, please try it out (see https://github.com/rust-lang/rustup/pull/3653#issuecomment-1936976011) and let us know what you think about it! :)

**Comment by smoelius** [comment]
> The beta release seems to be available right now. If you don't mind, please try it out and let us know what you think about it! :)

@rami3l Please forgive the naive question, but how do I try out the beta?

**Comment by rami3l** [comment]
> @rami3l Please forgive the naive question, but how do I try out the beta?

@smoelius No worries! Please refer to https://github.com/rust-lang/rustup/pull/3653#issuecomment-1936976011.

**Comment by smoelius** [comment]
:facepalm: Thanks, @rami3l.
